donkeycar/parts/object_detector/od_protocol.py

```diff
@@ -0,0 +1,283 @@
+            self.newimageready = True   
+    
+    def reset_run_counter(self):
+        self.run_counter = 0
+ 
+    def detect(self):  
+        self.newimageready = False
+        self.detector_ready = False
+        self.bbox = None
+        self.score = 0
+        self.name = None
+        self.position = 0.0
+        if self.image is not None:
+            self.bbox, self.score, self.name = self.detector.detect(self.image)
+        if self.bbox != None:
+            self.position = ((self.bbox.origin_x + (self.bbox.width / 2)) - self.img_center) / self.img_center  
+        self.detector_ready = True
+    
+    def update(self):
+        while self.on:
+           if self.newimageready:
+               self.detect()
+ 
+    def manage(self, angle, throttle):
+        # template routine
+        return angle, throttle
+
+    def run_threaded(self, angle, throttle, image):
+        self.ready()
+        angle, throttle = self.manage(angle, throttle)
+        self.next_image(image)
+        if self.detector_image is None:
+            self.detector_image = image
+        return angle, throttle, self.detector_image   
+    
+    def run(self, angle, throttle, image):
+        self.run_counter += 1
+        if self.run_counter >= self.run_trigger:
+            self.image = image
+            self.detect()
+            self.ready()
+            angle, throttle = self.manage(angle, throttle)
+            return angle, throttle, self.detector_image
+        else:
+            return angle, throttle, image           
+
+    def shutdown(self):
+        logger.info(f'Detector - average detection time {self.detector.average_perf():5.3f}')
+        self.on = False
+        
+class Stop_Manager():
+    # Stop states
+    stop_states = ('idle', 'start', 'back1', 'forward', 'back2')
+    IDLE = 0
+    INITIATE = 1
+    POS_ONE = 3
+    NEG_ONE = 2
+    NEG_TWO = 4
+    THROTTLE_INC = 0.2
+    
+    def __init__(self):
+        self.stop_state = self.IDLE
+        self.last_throttle = 0.0
+    
+    def stop(self):
+        if self.stop_state == self.IDLE:
+            self.stop_state = self.INITIATE
+    
+    def is_idle(self):
+        return self.stop_state == self.IDLE
+    
+    def throttle(self):
+#         if self.stop_state == self.IDLE:
+#             pass
+        throttle = 0.0
+        if self.stop_state == self.INITIATE:
+            self.stop_state = self.NEG_ONE
+            throttle = -1.0
+        elif self.stop_state == self.NEG_ONE:
+            self.stop_state = self.POS_ONE
+            throttle = 0.0
+        elif self.stop_state == self.POS_ONE:
+            self.stop_state = self.NEG_TWO
+            throttle = -1.0
+        elif self.stop_state == self.NEG_TWO:
+            throttle = self.last_throttle + self.THROTTLE_INC
+            if throttle >= 0.0:
+                throttle = 0.0
+                self.stop_state = self.IDLE
+        self.last_throttle = throttle
+        return throttle
+
+'''
```

Comment:
Can you add this as a docstring to the `__init__` method rather, same question for the class below. It's simply the convention we have been following in DC.

Reply:
decided to remove it instead